src/cliff_vesting/src/main.nr

```diff
@@ -0,0 +1,141 @@
+use dep::aztec::macros::aztec;
+
+#[aztec]
+pub contract CliffVesting {
+    use dep::aztec::{
+        encrypted_logs::encrypted_note_emission::encode_and_encrypt_note,
+        note::note_getter_options::NoteGetterOptions,
+        macros::{functions::{initializer, private}, storage::storage},
+        note::{Note, NoteHeader},
+    };
+    use dep::aztec::prelude::{AztecAddress, PrivateImmutable};
+
+    use dep::address_note::address_note::AddressNote;
+    use dep::token::Token;
+
+    use crate::escrow_contract;
+
+    #[storage]
+    struct Storage<Context> {
+        escrows: PrivateSet<CliffVestingNote, Context>,
+    }
+
+    #[note]
+    pub struct CliffVestingNote {
+        sender: AztecAddress,
+        receiver: AztecAddress,
+        escrow: AztecAddress,
+        vesting: U128,
+        randomness: Field
+    }
+
+    impl CliffVestingNote {
+        pub fn new(
+            sender: AztecAddress,
+            receiver: AztecAddress,
+            escrow: AztecAddress,
+            vesting: U128,
+            randomness: Field,
+        ) -> Self {
+            CliffVestingNote { 
+                sender,
+                receiver, 
+                escrow,
+                vesting,
+                randomness,
+                header: NoteHeader::empty() 
+            }
+        }
+    }
+
+    pub fn filter_note_escrow(
+        notes: [Option<CliffVestingNote>; MAX_NOTE_HASH_READ_REQUESTS_PER_CALL],
+        escrow: AztecAddress,
+    ) -> [Option<CliffVestingNote>; MAX_NOTE_HASH_READ_REQUESTS_PER_CALL] {
+        let mut selected = [Option::none(); MAX_NOTE_HASH_READ_REQUESTS_PER_CALL];
+
+        for i in 0..notes.len() {
+            if notes[i].is_some() {
+                let note = notes[i].unwrap_unchecked();
+                if note.escrow == escrow {
+                    selected[0] = Option::some(note);
+                    selected
+                }
+            }
+        }
+
+        assert(false, "No cliff vesting found for escrow");
+    }
+
+    #[private]
+    fn create_cliff_vesting(escrow: AztecAddress, receiver: AztecAddress, vesting: U128) {
+        let escrow_owner = Escrow::at(escrow).owner();
+        assert(escrow_owner == context.this_address(), "Cliff vesting is not owner");
+
+        let sender = context.msg_sender();
+        let note = CliffVestingNote::new(
+            sender, 
+            receiver, 
+            escrow, 
+            vesting, 
+            context.randomness()
+        );
+
+        // Do we need to check if there is already a cliff vesting created for this escrow?
+        let note_insertion = storage.escrows.insert(&mut context, note);
```

Comment:
we cannot check it like this, because somebody could "forget" a previous note using that escrow and it will pass, we need to enforce it with nullifiers, if we wanna keep 1-vest:1-escrow, we need to emit a nullifier `H(Escrow)` and avoid it from being used again (from this contract, which is good, bc this contract owns the escrow)